src/VSIX/NpgsqlConnectionEquivalencyComparer.cs

```diff
@@ -0,0 +1,24 @@
+﻿using Microsoft.VisualStudio.Data.Services.SupportEntities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Npgsql.VSIX
+{
+    class NpgsqlConnectionEquivalencyComparer : IVsDataConnectionEquivalencyComparer
```

Comment:
Microsoft.VisualStudio.Data.Framework provides `DataConnectionEquivalencyComparer`, which is an implementation of `IVsDataConnectionEquivalencyComparer`. I decompiled it with dotpeek to understand what's going on - take a look.It seems to parse the strings with the `IVsDataConnectionProperties` from the provider (so `NpgsqlConnectionProperties` in our case), and then delegates to an `AreEquivalent` overload which calls `ToString()` on the `IVsDataConnectionProperties`. I think the idea is to let the provider's `IVsDataConnectionProperties` do any custom added logic rather than going to the `DbConnectionStringBuilder` directly.

Note that `DataConnectionEquivalencyComparer` compares the `ToString()`s, so 2 strings are only considered equivalent if they match completely; this is probably because it doesn't know anything about any specific connection string properties. What you did (compare host, port, database, username) is probably the right thing to do

So it seems like it may be more correct for `NpgsqlConnectionEquivalencyComparer` to extend `DataConnectionEquivalencyComparer` and override the `AreEquivalent` that accepts `IVsDataConnectionProperties`, what do you think?